devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy

```diff
@@ -42,64 +43,140 @@ metadata {
+def getLEVEL_STEP_COMMAND() { 0x02 }
+def getLEVEL_STOP_COMMAND() { 0x03 }
+def getLEVEL_MOVE_LEVEL_ONOFF_COMMAND() { 0x04 }
+def getLEVEL_MOVE_ONOFF_COMMAND() { 0x05 }
+def getLEVEL_STEP_ONOFF_COMMAND() { 0x06 }
+def getLEVEL_STOP_ONOFF_COMMAND() { 0x07 }
+def getLEVEL_DIRECTION_UP() { "00" }
+def getLEVEL_DIRECTION_DOWN() { "01" }
+
+def getUINT8_STR() { "20" }
+
+
+def isIkeaDimmer() {
+	device.getDataValue("model") == "TRADFRI wireless dimmer"
+}
+def isAuroraRemote() {
+	device.getDataValue("model") == "Remote50AU"
+}
+def isMullerLichtTint() {
+	device.getDataValue("model") == "ZBT-DIMController-D0800"
+}
 
 // Parse incoming device messages to generate events
 def parse(String description) {
 	log.debug "description is $description"
+	def results = []
 
 	def event = zigbee.getEvent(description)
 	if (event) {
 		if (event.name=="level" && event.value==0) {}
 		else {
-			sendEvent(event)
+			results << createEvent(event)
 		}
 	} else {
 		def descMap = zigbee.parseDescriptionAsMap(description)
-		if (descMap && descMap.clusterInt == 0x0006) {
-			if (descMap.commandInt == 0x01 || descMap.commandInt == 0x00) {
+
+		if (descMap && descMap.clusterInt == zigbee.ONOFF_CLUSTER) {
+			if (descMap.commandInt == ONOFF_ON_COMMAND || descMap.commandInt == ONOFF_OFF_COMMAND) {
 				if (device.currentValue("level") == 0) {
-					sendEvent(name: "level", value: STEP)
+					results << createEvent(name: "level", value: STEP)
+				}
+
+				if (isAuroraRemote() || isMullerLichtTint()) { // Devices that have one power button
+					results << createEvent(name: "switch", value: device.currentValue("switch") == "on" ? "off" : "on")
+				} else {
+					results << createEvent(name: "switch", value: descMap.commandInt == ONOFF_OFF_COMMAND ? "off" : "on")
 				}
-				sendEvent(name: "switch", value: device.currentValue("switch") == "on" ? "off" : "on")
 			}
-		} else if (descMap && descMap.clusterInt == 0x0008) {
-			def currentLevel = device.currentValue("level") as Integer ?: 0
-			if (descMap.commandInt == 0x02) {
-				def value = Math.min(currentLevel + STEP, 100)
-				log.debug "move to ${descMap.data}"
+		} else if (descMap && descMap.clusterInt == zigbee.LEVEL_CONTROL_CLUSTER) {
+			if (descMap.commandInt == LEVEL_STEP_COMMAND) {
+				results += handleStepEvent(descMap)
+			} else if (isIkeaDimmer() && (descMap.commandInt == LEVEL_MOVE_COMMAND || descMap.commandInt == LEVEL_MOVE_ONOFF_COMMAND)) {
+				// For the IKEA Dimmer we are going to treat Level Move and Level Move with On/Off as Level Step
+				results += handleStepEvent(descMap)
+			} else if (descMap.commandInt == LEVEL_MOVE_COMMAND) {
+				results << createEvent(name: "level", value: descMap.data[0] == LEVEL_DIRECTION_UP ? 100 : STEP)
+				results << createEvent(name: "switch", value: "on")
+
+				log.debug "step to ${descMap.data}"
+			} else if (descMap.commandInt == LEVEL_STOP_COMMAND || descMap.commandInt == LEVEL_STOP_ONOFF_COMMAND) {
+				log.debug "stop move"
```

Comment:
If we don't need to create an event in this case I think it would be good to explicitly state that in a comment